RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -566,69 +567,52 @@ void L2TauNNProducer::fillCaloRecHits(tensorflow::Tensor& cellGridMatrix,
   }
 }
 
-std::vector<int> L2TauNNProducer::selectGoodVertices(const ZVertexSoA& patavtx_soa,
-                                                     const pixelTrack::TrackSoA& patatracks_tsoa,
-                                                     const std::vector<int>& TrackGood) {
-  auto maxTracks = patatracks_tsoa.stride();
+void L2TauNNProducer::selectGoodTracksAndVertices(const ZVertexSoA& patavtx_soa,
+                                                  const pixelTrack::TrackSoA& patatracks_tsoa,
+                                                  std::vector<int>& TrkGood,
+                                                  std::vector<int>& VtxGood) {
+  const auto maxTracks = patatracks_tsoa.stride();
   const int nv = patavtx_soa.nvFinal;
-  std::vector<int> VtxGood;
-  if (nv == 0)
-    return VtxGood;
-  VtxGood.reserve(nv);
-
-  std::vector<double> maxChi2_;
-  std::vector<double> pTSquaredSum(nv);
-
-  for (int j = nv - 1; j >= 0; --j) {
-    std::vector<int> trk_ass_to_vtx;
-    auto vtx_idx = patavtx_soa.sortInd[j];
-    assert(vtx_idx < nv);
-    for (int trk_idx = 0; trk_idx < maxTracks; trk_idx++) {
-      int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
-      if (vtx_ass_to_track == int16_t(vtx_idx))
-        trk_ass_to_vtx.push_back(trk_idx);
-    }
-    auto nt = trk_ass_to_vtx.size();
-    if (nt == 0) {
-      continue;
-    }
-    if (nt < 2) {
-      trk_ass_to_vtx.clear();
-      continue;
+  auto const* quality = patatracks_tsoa.qualityData();
+
+  // No need to sort either as the algorithms is just using the max (not even the location, just the max value of pt2sum).
+  std::vector<double> pTSquaredSum(nv, 0);
+  std::vector<int> nTrkAssociated(nv, 0);
+
+  for (int32_t trk_idx = 0; trk_idx < maxTracks; ++trk_idx) {
+    auto nHits = patatracks_tsoa.nHits(trk_idx);
+    if (nHits == 0) {
+      break;
     }
-    for (const auto& trk_idx : trk_ass_to_vtx) {
-      int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
-      if (vtx_ass_to_track != vtx_idx)
-        continue;
+    int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
+    if (vtx_ass_to_track >= 0 && vtx_ass_to_track <= nv) {
       double patatrackPt = patatracks_tsoa.pt[trk_idx];
```

Comment:
```suggestion
      auto patatrackPt = patatracks_tsoa.pt[trk_idx];
```
`auto` would give `float` rather than `double`, afaiu. Or was `double` intentional?